src/libstd/io/mod.rs

```diff
@@ -0,0 +1,928 @@
+    ///
+    /// It is possible to seek beyond the end of an object, but is an error to
+    /// seek before byte 0.
+    End(i64),
+
+    /// Set the offset to the current position plus the specified number of
+    /// bytes.
+    ///
+    /// It is possible to seek beyond the end of an object, but is an error to
+    /// seek before byte 0.
+    Current(i64),
+}
+
+fn read_until<R: BufRead + ?Sized>(r: &mut R, delim: u8, buf: &mut Vec<u8>)
+                                   -> Result<()> {
+    loop {
+        let (done, used) = {
+            let available = match r.fill_buf() {
+                Ok(n) if n.len() == 0 => return Ok(()),
+                Ok(n) => n,
+                Err(ref e) if e.kind() == ErrorKind::Interrupted => continue,
+                Err(e) => return Err(e)
+            };
+            match available.position_elem(&delim) {
+                Some(i) => {
+                    buf.push_all(&available[..i + 1]);
+                    (true, i + 1)
+                }
+                None => {
+                    buf.push_all(available);
+                    (false, available.len())
+                }
+            }
+        };
+        r.consume(used);
+        if done {
+            return Ok(());
+        }
+    }
+}
+
+/// A Buffer is a type of reader which has some form of internal buffering to
+/// allow certain kinds of reading operations to be more optimized than others.
+///
+/// This type extends the `Read` trait with a few methods that are not
+/// possible to reasonably implement with purely a read interface.
+pub trait BufRead: Read {
+    /// Fills the internal buffer of this object, returning the buffer contents.
+    ///
+    /// Note that none of the contents will be "read" in the sense that later
+    /// calling `read` may return the same contents.
+    ///
+    /// The `consume` function must be called with the number of bytes that are
+    /// consumed from this buffer returned to ensure that the bytes are never
+    /// returned twice.
+    ///
+    /// An empty buffer returned indicates that the stream has reached EOF.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an I/O error if the underlying reader was
+    /// read, but returned an error.
+    fn fill_buf(&mut self) -> Result<&[u8]>;
+
+    /// Tells this buffer that `amt` bytes have been consumed from the buffer,
+    /// so they should no longer be returned in calls to `read`.
+    fn consume(&mut self, amt: usize);
+
+    /// Read all bytes until the delimiter `byte` is reached.
+    ///
+    /// This function will consume all bytes in the underlying stream until the
+    /// specified delimiter is found. All bytes read are appended to the
+    /// specified buffer `buf`.
+    ///
+    /// The delimiter, if found, will be present in `buf` if `Ok` is returned.
+    ///
+    /// If this buffered reader is currently at EOF, then this function will not
+    /// place any more bytes into `buf` and will return `Ok(())`.
```

Comment:
And it's wrong for Readers where EOF does not indicate that later reads return EOF.